_posts/2020-10-22-how-to-update-to-aws-provider-v3.adoc

```diff
@@ -0,0 +1,191 @@
+tags: ["aws", "terraform", "terragrunt"]
+cloud: ["aws"]
+redirect_from: /static/guides/foundations/how-to-update-to-aws-provider-v3/
+---
+:page-type: guide
+:page-layout: post
+
+:toc:
+:toc-placement!:
+
+// GitHub specific settings. See https://gist.github.com/dcode/0cfbf2699a1fe9b46ff04c41721dda74 for details.
+ifdef::env-github[]
+:tip-caption: :bulb:
+:note-caption: :information_source:
+:important-caption: :heavy_exclamation_mark:
+:caution-caption: :fire:
+:warning-caption: :warning:
+toc::[]
+endif::[]
+
+== Intro
+
+This guide will walk you through how to update the https://gruntwork.io/reference-architecture/[Gruntwork Reference
+Architecture] and any code that depends on the
+https://gruntwork.io/infrastructure-as-code-library/[Gruntwork Infrastructure as Code Library] to the latest version
+(v3) of the Terraform AWS provider.
+
+=== What you'll learn in this guide
+
+This guide consists of three main sections:
+
+<<core_concepts>>::
+  An overview of version 3 of the AWS provider and why it is important to update your code for compatibility.
+
+<<deployment_walkthrough>>::
+  The steps you need to take to update your code relying on the Gruntwork Infrastructure as Code library and your
+  version of the Gruntwork Reference Architecture with compatibility with AWS provider v3.
+
+
+[[core_concepts]]
+== Core Concepts
+
+=== Background
+
+Version https://github.com/terraform-providers/terraform-provider-aws/releases/tag/v3.0.0[v3.0.0] of the Terraform AWS
+provider was released on July 30th 2020 with backwards incompatible updates. Following the provider release cycle,
+future releases from this point onward would only be compatible with 3.X. This means that if you wish to use any new
+resources or data sources that are released from this point onwards, you will need to be compatible with version 3 of
+the AWS provider to be able to pull those changes in.
+
+Since this is a major version change, there are a number of backwards incompatible changes that were introduced. In
+order to be able to use version 3, you need to update the resources and data sources with backwards incompatible
+changes.
+
+For your own code, you can follow
+https://registry.terraform.io/providers/hashicorp/aws/latest/docs/guides/version-3-upgrade[the official guide] on
+updating to AWS provider version 3. The rest of this guide will focus on updating references to Gruntwork modules and
+the Gruntwork Reference Architecture to be compatible with AWS provider version 3.
+
+
+[[deployment_walkthrough]]
+== Deployment walkthrough
+
+=== Updating references to Gruntwork Infrastructure as Code Library
+
+In order to take advantage of the Terraform AWS provider version 3, you need to update your references to the Gruntwork
+Infrastructure as Code Library to use a compatible version. The following table provides a summary of all the relevant
+Gruntwork AWS modules and the respective versions that are compatible with AWS provider version 3.
```

Comment:
You might want to mention here, and perhaps in the intro, that we (Gruntwork) did the hard work of testing and updating our whole library and Ref Arch to be compatible with AWS Provider 3.x. As a customer, you need to update to the proper versions of the Gruntwork library to pick up the fixes/changes we had to make, plus make some code and state changes in your own codebase.

Reply:
Also, after this paragraph, add an example of how to update. E.g., Look for code with `module { source = "github.com/gruntwork-io/xxx//yyy?ref=zzz" }`, bump `ref=zzz` to the corresponding version number below. And make sure to check out the release notes for those releases for migration guides!

Reply:
Reworded to include that we tested and updated the library: 259d8b08d0240ad8511b95945b3afc89ceb9440d

I didn't include the phrase "hard work", because IMHO, it feels a bit too pretentious and arrogant.

Also updated to link to "Updating" section of the "how to use Gruntwork" guide: ced5bfbe19a95becdb71daf97e128da9c1f9e0ae